packages/@lwc/template-compiler/src/parser/expression.ts

```diff
@@ -115,33 +115,37 @@ export function parseIdentifier(source: string): TemplateIdentifier | never {
 // Returns the immediate iterator parent if it exists.
 // Traverses up until it finds an element with forOf, or
 // a non-template element without a forOf.
-export function getForOfParent(element: IRElement): IRElement | null {
-    const parent = element.parent;
-    if (!parent) {
-        return null;
-    }
+export function getForOfParent(parentStack: IRElement[]): IRElement | null {
+    let size = parentStack.length;
+    let parent: IRElement | undefined = parentStack[--size];
+
+    while (parent) {
+        if (parent.forOf) {
+            return parent;
+        }
 
-    if (parent.forOf) {
-        return parent;
-    } else if (parent.tag.toLowerCase() === 'template') {
-        return getForOfParent(parent);
+        parent = parent.tag.toLowerCase() === 'template' ? parentStack[--size] : undefined;
```

Comment:
I am surprised that we have to do `String.prototype.toLowerCase()`. The tag should already be in lowercase.
I ran the test locally and it is not needed.